webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,239 @@
+  #syncSent?: string = undefined;
+  // syncStarted updates when recieving msg of {sync_id, complated: false}
+  #syncStarted?: string = undefined;
+  // syncStarted updates when recieving msg of {sync_id, complated: true}
+  #syncComplete?: string = undefined;
+  // List of messages recieved from server
+  #pendingMsg: Array<Record<string, StreamContent>> = [];
+
+  //callbacks
+  #onUpsert: (m: Record<string, StreamContent>) => void;
+  #onDelete: (s: Streamable, a: Array<number>) => void;
+  #isLoaded?: (ids: Array<string>) => void;
+
+  constructor(
+    wsUrl: string,
+    onUpsert: (m: Record<string, StreamContent>) => void,
+    onDelete: (s: Streamable, a: Array<number>) => void,
+    isLoaded?: (ids: Array<string>) => void,
+  ) {
+    this.#wsUrl = wsUrl;
+    this.#onUpsert = onUpsert;
+    this.#onDelete = onDelete;
+    this.#isLoaded = isLoaded;
+    this.#advance();
+  }
+
+  #connect(): WebSocket {
+    const ws = new WebSocket(this.#wsUrl);
+    ws.onopen = () => {
+      logger.info('Streaming websocket opened!');
+      this.#retries = 0;
+      this.#advance();
+    };
+    ws.onerror = (err) => {
+      // No need to do anything else becauses onerror will trigger onclose
+      logger.error('Streaming websocket errored: ', err);
+    };
+
+    ws.onclose = () => {
+      this.#syncSent = undefined;
+      const backoff = backoffs[this.#retries];
+      if (backoff === undefined) {
+        throw new Error('Websocket cannot reconnect!');
+      }
+      this.#timeout = dayjs().add(backoff, 'second');
+      this.#retries += 1;
+      logger.info(`#${this.#retries} of retries: in ${backoff}s`);
+      setTimeout(() => this.#advance(), backoff * 1000);
+    };
+
+    ws.onmessage = (event) => {
+      const msg = JSON.parse(event.data) as Record<string, StreamContent>;
+      this.#pendingMsg.push(msg);
+      this.#advance();
+    };
+
+    return ws;
+  }
+
+  #shouldSkip(newSub: SubscriptionGroup): boolean {
+    if (!this.#curSub) return false;
+    let same = true;
+    forEach(newSub, (val, k) => {
+      if (!this.#curSub?.[k as Streamable]?.spec.equals(val?.spec)) same = false;
+    });
+    return same;
+  }
+
+  #sendSpec(newSub: SubscriptionGroup): void {
+    // Skip current sub and move to next only if sub has already been sent
+    if (this.#shouldSkip(newSub) && this.#syncSent) {
+      this.#advance();
+      return;
+    }
+
+    const newSpecWithCache: Partial<Record<Streamable, SubscriptionWithCache>> = {};
+    this.#numSyncs += 1;
+    const sync_id = this.#numSyncs.toString();
+
+    const wholeSub = this.#curSub ? { ...this.#curSub, ...newSub } : newSub;
+
+    const payload = reduce(
+      wholeSub,
+      (payload, ent) => {
+        if (!ent) return payload;
+        const k = ent.spec.id();
+        const curSpec = this.#curSub?.[k];
+        const keyCache =
+          curSpec && ent.spec.contains(curSpec.spec) ? curSpec.keyCache.copy() : new KeyCache();
```

Comment:
question: why not always keep the current keycache?  If there is any overlap between the old subscription and the new one, you'll almost certainly save bandwidth by reusing the objects you already have stored where the overlap exists.

And since it's hard to calculate the overlap here... you could just let the server do it for you, right?  That's why the python client keeps its keycaches forever, basically.

Reply:
I agree, we can have `XXX_deleted` to handle everything